src/Cli/dotnet/ToolPackage/ToolPackageDownloader.cs

```diff
@@ -0,0 +1,311 @@
+﻿// Copyright (c) .NET Foundation and contributors. All rights reserved.
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+
+using System.CommandLine;
+using System.Reflection;
+using Microsoft.DotNet.Cli.NuGetPackageDownloader;
+using Microsoft.DotNet.Cli.Utils;
+using Microsoft.DotNet.ToolPackage;
+using Microsoft.DotNet.Tools;
+using Microsoft.Extensions.EnvironmentAbstractions;
+using NuGet.Client;
+using NuGet.Common;
+using NuGet.ContentModel;
+using NuGet.Frameworks;
+using NuGet.LibraryModel;
+using NuGet.Packaging;
+using NuGet.Packaging.Core;
+using NuGet.ProjectModel;
+using NuGet.Repositories;
+using NuGet.RuntimeModel;
+using NuGet.Versioning;
+using System;
+
+
+namespace Microsoft.DotNet.Cli.ToolPackage
+{
+    internal class ToolPackageDownloader : IToolPackageDownloader
+    {
+        private INuGetPackageDownloader _nugetPackageDownloader;
+        private readonly IToolPackageStore _toolPackageStore;
+
+        // The directory that the tool will be downloaded to
+        protected DirectoryPath _toolDownloadDir;
+
+        // The directory that the tool package is returned 
+        protected DirectoryPath _toolReturnPackageDirectory;
+
+        // The directory that the tool asset file is returned
+        protected DirectoryPath _toolReturnJsonParentDirectory;
+
+        // The directory that global tools first downloaded
+        // example: C:\Users\username\.dotnet\tools\.store\.stage\tempFolder
+        protected readonly DirectoryPath _globalToolStageDir;
+
+        // The directory that local tools first downloaded
+        // example: C:\Users\username\.nuget\package
+        protected readonly DirectoryPath _localToolDownloadDir;
+
+        // The directory that local tools' asset files located
+        // example: C:\Users\username\AppData\Local\Temp\tempFolder
+        protected readonly DirectoryPath _localToolAssetDir;
+
+        public static readonly CliOption<VerbosityOptions> VerbosityOption = CommonOptions.VerbosityOption;
+
+        protected readonly string _runtimeJsonPath;
+
+        public ToolPackageDownloader(
+            IToolPackageStore store,
+            string runtimeJsonPathTest = null
+            // string localToolPath = null
+        )
+        {
+            _toolPackageStore = store ?? throw new ArgumentNullException(nameof(store)); ;
+            _globalToolStageDir = _toolPackageStore.GetRandomStagingDirectory();
+            /*string localToolDownloadDir = localToolPath
+                ?? (Environment.GetEnvironmentVariable("NUGET_PACKAGES"))
+                ?? (Path.Combine(Environment.GetFolderPath(Environment.SpecialFolder.UserProfile), "nuget", "package"));*/
+            string localToolDownloadDir = (Environment.GetEnvironmentVariable("NUGET_PACKAGES"))
+                ?? (Path.Combine(Environment.GetFolderPath(Environment.SpecialFolder.UserProfile), "nuget", "package"));
```

Comment:
Are you avoiding the `globalPackageDirectory` value in nuget.config intentionally?  Currently with the restore-based tool install, it will consider nuget.config, meaning this would be a change in behaviour. Maybe that's acceptable, I don't know, but if I

I previously provided a comment with the APIs needed to get the global packages folder in the same way that NuGet does for all our other experiences: https://github.com/dotnet/sdk/pull/33835#discussion_r1295679650

Reply:
Yes, please use the NuGet API to get the global packages folder.  The settings should be created the same way as they are created in NuGetPackageDownloader (they are where the feeds and package source mapping are defined).